### ISSUE TITLE
Apollo federation reference for typename

### ISSUE THREAD

**Issue by renanpupin** [hayes/pothos#1146]
On apollo federation, is possible to return a typename to delegate the resolve method of a field to another subgraph.
For example, on customers graph, it's possible to resolve the field "order" by delegating the typename to subgraph, in this case the order graph:
` return { __typename: "Order", customerId: customer._id };`

In the example below, customer can resolve this orders by delegating the ref to the graph that resolves the typename "Orders", than this graph can provide fields to another type by referencing it's typename.
```
const customerResolvers = {
  Customer: {
    __resolveReference(object) {
      //customer can resolve itself 
      return customers.find((customer) => customer._id === object._id);
    },
    Orders(customer) {
      //customer delegates the resolution of orders to another graph
      return { __typename: "Order", customerId: customer._id };
    }
  }
};
const orderResolvers = {
  Order: {
    __resolveReference(object) {
      //order can resolve itself 
      return orders.find((order) => order.customerId === object.customerId);
    }
  }
};
```
In this case, both types are in the sabe subgraph, they are not external refs. Currently apollo schema first SDL supports this kinds of operation.
The question is, how can we do it using pothos? Documentation doesn't provide an example of this use case.

**Comment by JCEsmeraldo** [comment]
I have the same issue!

**Comment by hayes** [comment]
Can you share what the SDL version of this looks like?  I can't really understand what this is supposed to look like with just the resolvers.

**Comment by renanpupin** [comment]
We can use this example to propagate the __typename of the graph entity and delegate it to be resolved by some of the graphs:
https://gist.github.com/Oxyaction/bbe9f4915d00e98344574a12e22ac9ee#file-index-ts

There is also a blog post about this:
https://medium.com/@oxyaction/apollo-federation-v2-by-example-72bba5bd5cdb

**Comment by hayes** [comment]
In the blog post example, Product is just defined as an entity in both SubGraphs.  They don't explain it that way, but federation v2 allows entities to be defined in multiple sub-graphs, so long as their keys are compatible.  An entity is just an object type with a key directive.  I don't think the article is correct that your resolver needs to a return a __typename field unless your field is an interface/union.   What actually happens there is that the gateway will resolve the id + __typename fields in it's query to you service.


Eg it might query reviews with:

```gql
query {
  reviews { id, product { __typename, id } }
}
```

Then it can get more details through the entities query

```gql
query {
  entities(keys: [{ __typename: "Product", id: 123 }]) {
      ...on Product {
      id
      __typename
      name
      price
    }
}
```


Basically, you would define Product in the reviews schema like:


```ts
const Product = builder.objectRef<{ id: string }>(Product').implement({
  fields: (t) => ({
    id: t.exposeID('id'),
    reviews: t.field({
      type: [Review],
      resolve: (product) => find reviews(product.id)
  }),
});

builder.asEntity(Product, {
  key: builder.selection<{ id: string }>('id'),
  resolveReference: (product) => product
});

**Comment by renanpupin** [comment]
Can you elaborate please what this find function should come from since reviews is declared in another subgraph?
`resolve: (product) => find reviews(product.id)`

**Comment by renanpupin** [comment]
I tried to apply the example from the federation plugin tests ([link](https://github.com/hayes/pothos/tree/f867a9f18cc77fb768d64eb300ceecec51de7c67/packages/plugin-federation/tests/example)).
What I can see is that pothos complains about having duplicate types, even if I define the external ref on the it's typings.
<img width="1010" alt="Captura de Tela 2024-02-03 às 12 18 29" src="https://github.com/hayes/pothos/assets/8743530/3d123af6-671f-4ea1-b603-aa2bcf732bdd">

<img width="679" alt="Captura de Tela 2024-02-03 às 12 17 33" src="https://github.com/hayes/pothos/assets/8743530/7d9c03ba-69d0-4254-8d55-7807f2303059">

The fact is that I have Reviews and Product on the same subgraph, but for separation of concerns purposes, I wouldn't like to import reviews module services inside product modules services, instead, I wish to resolve it but just propagating the typename or some similar solution. (which is possible using the __typename strategy).

So I am missing something here?

**Comment by hayes** [comment]
> Can you please elaborate what this find function should come from since reviews is declared in another subgraph? `resolve: (product) => find reviews(product.id)`

This was showing what would be implemented in the Reviews SubGraph.   The products sub graph wouldn't have the reviews field defined

> The fact is that I have Reviews and Product on the same subgraph, but for separation of concerns purposes, I wouldn't like to import reviews module services inside product modules services, instead, I wish to resolve it but just propagating the typename or some similar solution. (which is possible using the __typename strategy).

This doesn't make sense to me, if you aren't implementing these in different sub graphs, this wouldn't be related to federation at all.   Are you just trying to define the resolvers on separate files?  That's pretty straight forward if so, but I suspect I don't understand what your trying to do.  What does "modules services" mean? 

Fundamentally pothos is just providing a type-safe way to define the directives and resolveReference methods.   It doesn't change or implement the federation logic itself.  Anything that works in SDL based federation works exactly the same in Pothos, the only difference is where you define the resolvers and directives.  The execution all uses the exact same code under the hood

**Comment by hayes** [comment]
heres a working implementation of whats described in the article you linked: https://github.com/hayes/pothos/compare/mh--federation-product-review-example?expand=1

**Comment by hayes** [comment]
Closing this since I haven't heard back.  Hope you got this figured out, but feel free to re-open if not
